### ISSUE TITLE
`super()` should initialize the `this` binding with whatever value is returned from parent constructor

### ISSUE THREAD

**Issue by loganfsmyth** [babel/babel#2279]
The current implementation doesn't handle parent classes that return specific objects properly. The value returned from `[[Construct]]` should be used as the `this` binding value in the child constructor.

```
class Other {
    method(){
        console.log('other');
    }
}
class Base {
    constructor(){
        return new Other();
    }
    method(){
        console.log('base');
    }
}
class Sub extends Base {
    constructor(){
        super();
        this.method();
    }
}

var s = new Sub();
console.log('expected true, got ' + (s instanceof Other));
```

Seems like

```
var Sub = (function (_Base) {
    _inherits(Sub, _Base);

    function Sub() {
        _classCallCheck(this, Sub);

        _get(Object.getPrototypeOf(Sub.prototype), 'constructor', this).call(this);
        this.method();
    }

    return Sub;
})(Base);
```

should be more like

```
var Sub = (function (_Base) {
    _inherits(Sub, _Base);

    function Sub() {
        _classCallCheck(this, Sub);

        var _this = _get(Object.getPrototypeOf(Sub.prototype), 'constructor', this).call(this);
        if (_this === undefined) _this = this;
        _this.method();
        return _this;
    }

    return Sub;
})(Base);
```

**Comment by katrotz** [comment]
Might be the same issue. 
I am trying to extend the [SystemJS constructor](https://github.com/systemjs/systemjs/blob/master/docs/system-api.md#systemconstructor) like:

```
class MyFancySubclass extends System.constructor {
  constructor() {
    super();
    ...
  }
}
```
but in the end the `System.constructor` is not getting called when I do `super()`.

I have traced the potential issue in the babels `super` implementation:
The `inherits` babel runtime helper maps the constructor of the extended class on the `__proto__` of the subclass, but when the `super` is being invoked in the subclass constructor I see the following code:
`_get(Object.getPrototypeOf(MyFancySubclass.prototype), 'constructor', this).call(this);
`
which indeed will return the superClasses prototype, but not the superClasses constructor.

**Comment by katrotz** [comment]
I it might be related to https://github.com/babel/babel/issues/2048

**Comment by loganfsmyth** [comment]
@katrotz Yes this looks like related to that issue than this one. SystemJS is not properly settings `SystemJSLoader.prototype.constructor = SystemJSLoader` after this line: https://github.com/systemjs/systemjs/blob/af8f289a78fb92dd2c9d28881b62af1237980e1b/lib/proto.js#L12 so this something that would need to be fixed in SystemJS.

**Comment by sebmck** [comment]
Fixed in the `development` branch for 6.0.
